File: problems/leetcode/0900/951.cpp
// 951. Flip Equivalent Binary Trees
// https://leetcode.com/problems/flip-equivalent-binary-trees/

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
public:
  bool flipEquiv(TreeNode *root1, TreeNode *root2) {
    if (isBothNodeNull(root1, root2)) {
      return true;
    }

    if (isOnlyOneNodeNull(root1, root2)) {
      return false;
    }

    if (isNodeValDifferent(root1, root2)) {
      return false;
    }

    if (isEquivNode(root1->left, root2->left) &&
        isEquivNode(root1->right, root2->right)) {
      return flipEquiv(root1->left, root2->left) &&
             flipEquiv(root1->right, root2->right);
    }

    if (isEquivNode(root1->left, root2->right) &&
        isEquivNode(root1->right, root2->left)) {
      return flipEquiv(root1->left, root2->right) &&
             flipEquiv(root1->right, root2->left);
    }

    return false;
  }
  bool isBothNodeNull(TreeNode *node1, TreeNode *node2) {
    return !node1 && !node2;
  }
  bool isOnlyOneNodeNull(TreeNode *node1, TreeNode *node2) {
    return (!node1 && node2) || (node1 && !node2);
  }
  bool isNodeValDifferent(TreeNode *node1, TreeNode *node2) {
    return node1->val != node2->val;
  }
  bool isNodeValSame(TreeNode *node1, TreeNode *node2) {
    return node1 && node2 && node1->val == node2->val;
  }
  bool isEquivNode(TreeNode *node1, TreeNode *node2) {
    return isBothNodeNull(node1, node2) || isNodeValSame(node1, node2);
  }
};

File: problems/leetcode/0900/951.py
# 951. Flip Equivalent Binary Trees
# https://leetcode.com/problems/flip-equivalent-binary-trees/description

from typing import *

# 재귀적으로 확인
# 정방향으로 동일한지, 교차로 동일한지 모르기 때문에 두 케이스 다 검증하면 OK

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def flipEquiv(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        if not root1 and not root2:
            return True

        if (root1 and not root2) or (not root1 and root2):
            return False

        return root1.val == root2.val and (
            (self.flipEquiv(root1.left, root2.left) and self.flipEquiv(root1.right, root2.right)) or
            (self.flipEquiv(root1.left, root2.right) and self.flipEquiv(root1.right, root2.left))
        )
